L1Trigger/TrackFindingTracklet/src/Sector.cc

```diff
@@ -0,0 +1,486 @@
+#include "L1Trigger/TrackFindingTracklet/interface/CleanTrackMemory.h"
+
+#include "L1Trigger/TrackFindingTracklet/interface/VMRouter.h"
+#include "L1Trigger/TrackFindingTracklet/interface/TrackletEngine.h"
+#include "L1Trigger/TrackFindingTracklet/interface/TrackletEngineDisplaced.h"
+#include "L1Trigger/TrackFindingTracklet/interface/TripletEngine.h"
+#include "L1Trigger/TrackFindingTracklet/interface/TrackletCalculator.h"
+#include "L1Trigger/TrackFindingTracklet/interface/TrackletProcessor.h"
+#include "L1Trigger/TrackFindingTracklet/interface/TrackletCalculatorDisplaced.h"
+#include "L1Trigger/TrackFindingTracklet/interface/ProjectionRouter.h"
+#include "L1Trigger/TrackFindingTracklet/interface/MatchEngine.h"
+#include "L1Trigger/TrackFindingTracklet/interface/MatchCalculator.h"
+#include "L1Trigger/TrackFindingTracklet/interface/MatchProcessor.h"
+#include "L1Trigger/TrackFindingTracklet/interface/FitTrack.h"
+#include "L1Trigger/TrackFindingTracklet/interface/PurgeDuplicate.h"
+#include "L1Trigger/TrackFindingTracklet/interface/Util.h"
+
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+using namespace std;
+using namespace Trklet;
+
+Sector::Sector(unsigned int i, const Settings* settings, Globals* globals) : settings_(settings), globals_(globals) {
+  isector_ = i;
+  double dphi = 2 * M_PI / settings_->NSector();
+  double dphiHG = 0.5 * settings_->dphisectorHG() - M_PI / settings_->NSector();
+  phimin_ = isector_ * dphi - dphiHG;
+  phimax_ = phimin_ + dphi + 2 * dphiHG;
+  phimin_ -= M_PI / settings_->NSector();
+  phimax_ -= M_PI / settings_->NSector();
+  phimin_ = Trklet::phiRange(phimin_);
+  phimax_ = Trklet::phiRange(phimax_);
+  if (phimin_ > phimax_)
+    phimin_ -= 2 * M_PI;
+}
+
+Sector::~Sector() {
+  for (const auto& p : Processes_) {
+    ProcessBase* proc = p.second;
+    delete proc;
+  }
+  for (const auto& m : Memories_) {
+    MemoryBase* mem = m.second;
+    delete mem;
+  }
+}
+
+bool Sector::addStub(L1TStub stub, string dtc) {
+  bool add = false;
+
+  double phi = stub.phi();
+  double dphi = 0.5 * settings_->dphisectorHG() - M_PI / settings_->NSector();
+
+  std::map<string, std::vector<int> >& ILindex = globals_->ILindex();
+  std::vector<int>& tmp = ILindex[dtc];
+  if (tmp.size() == 0) {
+    for (unsigned int i = 0; i < IL_.size(); i++) {
+      if (IL_[i]->getName().find("_" + dtc) != string::npos) {
+        tmp.push_back(i);
+      }
+    }
+  }
+
+  if (((phi > phimin_ - dphi) && (phi < phimax_ + dphi)) ||
+      ((phi > 2 * M_PI + phimin_ - dphi) && (phi < 2 * M_PI + phimax_ + dphi))) {
+    Stub fpgastub(stub, settings_, phimin_, phimax_);
+    std::vector<int>& tmp = ILindex[dtc];
+    assert(tmp.size() != 0);
+    for (unsigned int i = 0; i < tmp.size(); i++) {
+      if (IL_[tmp[i]]->addStub(settings_, globals_, stub, fpgastub, dtc))
+        add = true;
+    }
+  }
+
+  return add;
+}
+
+void Sector::addMem(string memType, string memName) {
+  if (memType == "InputLink:") {
```

Comment:
The methods addMem and addProc are factories. If you mean a class to implement this I don't think there is a specific advantage to this. I did clean up the code a little to remove some repeated code.

Reply:
"factory" has a specific meaning in CMSSW. Helper algorithms that share a base class should be registered, and then can be produced automatically, rather than requiring an ever-growing if statement. See https://github.com/cms-sw/cmssw/blob/master/FWCore/PluginManager/interface/PluginFactory.h and search the code for various examples of factories using `EDM_REGISTER_PLUGINFACTORY` (trying to find a good twiki, but not sure if there is one)

Reply:
I think that's a valid suggestion for the long term development of the code, assuming it doesn't need to be compilable outside of CMSSW.

I don't see any mention of factories in the [CMS Coding Rules](https://cms-sw.github.io/cms_coding_rules.html), so I don't think this is a requirement.